modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -1038,350 +1071,107 @@ def _union_all(
-                if len(col_names) != 0:
-                    for frame in frames:
-                        table = frame.table
-                        idx_width = (
-                            0 if frame.index_cols is None else len(frame.index_cols)
-                        )
-                        table_cols = table.column_names[idx_width:]
-                        for col_name, table_col_name in zip(frame.columns, table_cols):
-                            if col_name in col_names:
-                                add_col_field(table, col_name, table_col_name)
-
-        if len(col_fields) == 0:
-            if ignore_index or len(frames) == 0:
-                idx = RangeIndex(0, sum(len(f.table) for f in frames))
-            else:
-                idx = frames[0].index.append([f.index for f in frames[1:]])
-            idx_cols = mangle_index_names(idx.names)
-            idx_df = pd.DataFrame(index=idx).reset_index()
-            union = pyarrow.Table.from_pandas(idx_df).rename_columns(idx_cols)
-        else:
-            # Process empty frames with non-empty index
-            for frame in frames:
-                if len(frame.table) == 0 and len(frame.index) != 0:
-                    if ignore_index:
-                        frame.index = pd.RangeIndex(0, len(frame.index))
-                    idx = frame.index
-                    frame.index_cols = mangle_index_names(idx.names)
-                    idx_df = pd.DataFrame(index=idx).reset_index()
-                    frame.table = pyarrow.Table.from_pandas(idx_df)
-                    frame.table = frame.table.rename_columns(frame.index_cols)
-
-            idx_cols = None
-            idx_table = None
-            idx_fields: typing.OrderedDict[str, pyarrow.Field] = OrderedDict()
-
-            if not ignore_index:
-                idx_cols = frames[0].index_cols
-                idx_equal = idx_cols is not None
-
-                if idx_equal:
-                    idx_width = len(idx_cols)
-                    idx_types = [frames[0].table.field(c).type for c in idx_cols]
-                    for frame in frames[1:]:
-                        table = frame.table
-                        frame_idx_cols = frame.index_cols
-                        if (
-                            (frame_idx_cols is None)
-                            or (len(frame_idx_cols) != idx_width)
-                            or any(
-                                idx_types[i] != table.field(frame_idx_cols[i]).type
-                                for i in range(idx_width)
-                            )
-                        ):
-                            idx_equal = False
-                            break
-
-                if idx_equal:
-                    idx_cols = list(
-                        concat_index_names([f.frame for f in frames]).keys()
+                table_columns = ColNameCodec.concat_index_names(frames)
+                index_cols = list(table_columns)
```

Comment:
This looks clearer
```suggestion
                index_cols = ColNameCodec.concat_index_names(frames)
                table_columns = list(index_cols)
```

Reply:
Sorry, but it's not correct. `table_columns` should be a dictionary, and `index_cols` - a list.

Reply:
`table_columns2types`?